### ISSUE TITLE
QueryBuilder combining 'where' conditions

### ISSUE THREAD

**Issue by CocaColaBear** [typeorm/typeorm#805]
Hi all. How i can implement this, without brackets?

```
let qb = this.repository.createQueryBuilder("account");

qb.where("account.email = :email", {email: email});
qb.andWhere("account.password = :password", {password: password});
qb.andWhere("(accountrole.expires < now() OR accountrole.expires IS NULL)"); // Brackets here
```

**Comment by pleerock** [comment]
only the way you wrote

**Comment by pleerock** [comment]
I mean you can't do what you want without brackets

**Comment by CocaColaBear** [comment]
It would be nice to have something like this:
qb.where("account.email = :email", {email: email});
qb.andWhere("account.password = :password", {password: password});
qb.andWhere(subQb => {
    subQb.where("accountrole.expires < now()");
    subQb.orWhere("accountrole.expires IS NULL");
});

**Comment by pleerock** [comment]
```ts
qb.andWhere(subQb => {
    subQb.where("accountrole.expires < now()");
    subQb.orWhere("accountrole.expires IS NULL");
});
```

this syntax is already taken for subqueries. But in your case you don't need subqueries

**Comment by emorris00** [comment]
> this syntax is already taken for subqueries. But in your case you don't need subqueries

That's too bad, I have actually written a change that adds that syntax for wheres and havings on my computer. I meant to add it as a pull request a while ago but I forgot.

**Comment by pleerock** [comment]
I don't think I would like to change exist subquery syntax...

```ts
qb.andWhere(new Brackets(sqb => {
    sqb.where("accountrole.expires < now()");
    sqb.orWhere("accountrole.expires IS NULL");
}));
```

**Comment by pleerock** [comment]
I implemented this feature as I described above:

```ts
        const users = await connection.createQueryBuilder(User, "user")
            .where("user.isAdmin = TRUE")
            .orWhere(new Brackets(qb => {
                qb.where("user.firstName = 'Timber'")
                    .andWhere("user.lastName = 'Saw'");
            }))
            .orWhere(new Brackets(qb => {
                qb.where("user.firstName = 'Alex'")
                    .andWhere("user.lastName = 'Messer'");
            }))
            .getMany();
```

Will be released in next version (0.1.0-alpha.36).

**Comment by iz-iznogood** [comment]
@pleerock Hi, Brackers is not exported. Is this intentional ?

**Comment by pleerock** [comment]
@iz-iznogood , no we need to add it, looks like I forgot

**Comment by LogansUA** [comment]
@pleerock can `Brackets` be used in `innerJoin` with subquery?

Here's my example
```typescript
const subQuery = this.createQueryBuilder("m2")
    .select("m2.id")
    .andWhere("m2.code = :code")
    .andWhere("m2.name = :name");

return this.createQueryBuilder("m")
    .innerJoin(
        () => subQuery,
        "existingModel",
        "existingModel.id = m.id",
    )
    .andWhere("m.id = :modelId")
    .setParameters({
        code,
        name,
        modelId: model.id,
    })
    .getOne();
```
And here's the query
```sql
SELECT
  "m"."id"         AS "m_id",
  "m"."code"       AS "m_code",
  "m"."name"       AS "m_name",
  "m"."enabled"    AS "m_enabled",
  "m"."created_at" AS "m_created_at",
  "m"."updated_at" AS "m_updated_at"
FROM
  "pss_models" "m"
  
  INNER JOIN
    SELECT
      "m2"."id" AS "m2_id"
    FROM
      "pss_models" "m2"
    WHERE
      "m2"."code" = $1
      AND "m2"."name" = $2
  "existingModel" ON existingModel.id = "m"."id"
WHERE
  "m"."id" = $3
```

As you can see, there's no brackets around my subquery in inner join.

Can you suggest me something?

**Comment by sushilbansal** [comment]
Hi @pleerock i need access to qb (Where Expression) created inside Brackets so that i can add more conditions to it later on. Is that possible?

Or in simpler terms i want to add more conditions to Brackets later on. Is there a way to do it.

**Comment by sushilbansal** [comment]
Resolved it by creatinga bracket inside bracket. Later on other And conditions work for both scenarios.

**Comment by cfi2017** [comment]
Wouldn't it be more convenient to make Brackets available like so:
```typescript
queryBuilder
  .where(/*condition*/)
  .andWhere(/*condition*/)
  .andBrackets(qb => 
    qb.where(/*condition*/)
      .orWhere(/*condition*/)
  )
```

This would remove the need to initialise a class.  
`andBrackets(callback: (qb) => any)` and `orBrackets(callback: (qb) => any)`  
could then be shorthands for  
`andWhere(new Brackets(callback: (qb) => any))` and `orWhere(new Brackets(callback: (qb) => any))`

**Comment by osnersanchez** [comment]
Is this in some documentation?
